Unit Test coverage

mkalcok opened the pull request:
Main goal of this change is to add 100% coverage of easyrsa charm with unit tests. There are also few small changes along for the ride.

- fixed typos and non-functional changes in the main code (I'll add comments to each of them)
- Fixed problem that some tests in `test_actions.py` were not passing on `python3.5` due to usage of `Mock.assert_called`
- Modified github workflow to also run unit tests and linters on opened PR, not just pushes
- Renamed tox environment `py3` to `unit` to bring it more in line with other charms.

mkalcok committed:
Split UTs for `restore` action and add 100% coverage for actions.py

mkalcok committed:
Fix typos and small non-functional problems

mkalcok committed:
Add unit tests for easyrsa layer

mkalcok committed:
Remove usage of `Mock.assert_called`.

It's unsupported in python3.5 and it also leads to slopier tests where proper arguments are not checked.

mkalcok committed:
Run linters and unit tests on each opened PR.

Also renamed tox env "py3" to "unit" to bring it more in line with other charms.

mkalcok commented on reactive/easyrsa.py:
```diff
@@ -1,3 +1,4 @@
+import ipaddress
```
This package was previously imported inside of a function

mkalcok commented on reactive/easyrsa.py:
```diff
@@ -371,8 +372,6 @@ def create_server_certificate(cn, san_list, name=None):
     if name is None:
         name = 'server'
-    server_cert = None
-    server_key = None
```
`server_cert` and `server_key` variables do not need to be initialised here because they are later assigned values (non-conditionally)

mkalcok commented on reactive/easyrsa.py:
```diff
@@ -414,8 +413,6 @@ def create_server_certificate(cn, san_list, name=None):
 def create_client_certificate(name='client'):
     '''Return a newly created client certificate and client key, by name.'''
-    client_cert = None
-    client_key = None
```
`client_cert` and `client_key` variables do not need to be initialised here because they are later assigned values (non-conditionally)

mkalcok commented on reactive/easyrsa.py:
```diff
@@ -449,8 +446,9 @@ def install_ca(certificate_authority):
 
 
-def get_sans(address_list=[]):
+def get_sans(address_list=None):
```
default value for an argument should not be a mutable object

mkalcok committed:
Remove usage of `shlex.split` in tests.

This caused exception when tests ran in some environments.

mkalcok commented:
🤔 CI runs raise exceptions when `shlex.split()` is called (whether in tests or in tested code). I'm not experiencing the same behaviour on my system, I'll have to find to reproduce it and I'll fix it.

rgildein commented:
> thinking CI runs raise exceptions when `shlex.split()` is called (whether in tests or in tested code). I'm not experiencing the same behaviour on my system, I'll have to find to reproduce it and I'll fix it.

I tried it I got similar [issue](https://pastebin.ubuntu.com/p/3rsFVKnYxT/).

rgildein commented on .github/workflows/tox.yaml:
```diff
@@ -1,6 +1,11 @@
 name: Run tests with Tox
 
-on: [push]
+on:
```
Maybe we could add Python 3.9 into matrix?

mkalcok replied:
I was not sure about what python version are really expected to be supported so I left it as it was. Maybe someone fro k8s team can comment on this.

auria replied:
The recently released Ubuntu version supports 3.9. In 22.04, a new LTS version will be released and it will probably support the next version.

See https://packages.ubuntu.com/search?keywords=python3&searchon=names&exact=1&suite=all&section=all

rgildein commented on tests/test_actions.py:
```diff
@@ -479,48 +522,137 @@ def test_restore_action(self):
-        tls_server_relation = tls_certificate_relation('tls_server',
-                                                       'server')
-        tls_provider.all_requests = [tls_server_relation]
+        # builtin 'open()' function will raise FileNotFoundError, which acts
```
When I tried to run:
```python
    with self.func_call_arguments(name='not-valid-backup.tar.gz'):
        self.call_action()
        self.assert_function_fail_msg("File not found")
```
I fount out that `function_fail` was called with arguments:
```python
[call('Action restore failed: Unrecognized certificate request type "application".')]
``` 
Is this ok?

mkalcok replied:
Hmm..when I change filename and add the `assert_function_fail_msg("FIle not found")` I get
```
E           AssertionError: expected call not found.
E           Expected: mock('Action restore failed: File not found')
E           Actual: not called.
```
Which is expected, action should not fail in this UT. Changing name of the "backup file" in `func_call_arguments` does not really matter, to simulate missing backup file, you'd need to set `os.path.isfile` return value to `False`, but there's already test for that (`test_missing_backup_file`)

rgildein replied:
> Hmm..when I change filename and add the `assert_function_fail_msg("FIle not found")` I get
> 
> ```
> E           AssertionError: expected call not found.
> E           Expected: mock('Action restore failed: File not found')
> E           Actual: not called.
> ```

Same here, however I printed all calls and there was only `[call('Action restore failed: Unrecognized certificate request type "application".')]`, which seems little bit odd, because I expected something like `file not found`.

> Which is expected, action should not fail in this UT. Changing name of the "backup file" in `func_call_arguments` does not really matter, to simulate missing backup file, you'd need to set `os.path.isfile` return value to `False`, but there's already test for that (`test_missing_backup_file`)

Oh well, I get it. Thanks for the explanation.

mkalcok replied:
Are you sure that the call `'Action restore failed: Unrecognized certificate request type "application".'` was in this test and not the one below, `test_restore_action_unknown_cert_type`?
This test really should not have any calls to `function_fail` and I can not reproduce it.

rgildein replied:
Actually, it came from `tests/test_actions.py::RestoreActionTests::test_restore_action_unknown_cert_type`, I just misled it.

mkalcok replied:
Ok, good, that's expected then.

rgildein reviewed (commented):
one question about raising the FileNotFoundError

mkalcok committed:
Revert "Remove usage of `shlex.split` in tests."

This reverts commit db22c55d05e3a348c09ce14626db3117ef9511ab.

mkalcok committed:
Fix problems when patching `os.path.islink`

mkalcok commented:
Ok, so the weird problems with `shlex.split()` were fixed. (For some more context, same issue was described [here](https://github.com/nedbat/coveragepy/issues/518)).
It was not actually caused by `shlex.split`, but by `coverage` which had problems when `os.path.islink` function was mocked. Solution to this is patching `islink` only in the tested module. However that's what I thought I was doing in the first place.

`easyrsa.py` imported only `os` package and then used `os.path.islink`, so I originally patched it like this
```python
is_link_patch = patch.object(easyrsa.os.path, "islink")
```
thinking that this will limit scope of the patch to `islink` object imported in the `easyrsa.py` module. However that was not the case and `islink` was patched globally, causing `coverage` to fail in certain scenarios.

The only way to avoid this was to change `easyrsa.py` to import directly `from os.path import islink` and then patch
```python
is_link_patch = patch.object(easyrsa, "islink")
``` 

This is not ideal solution because it forces the main code to change in order for tests to pass, but I did not found other way around it.

Note: The original problem did not appear in my dev environment because I was using `python3.10`, once I used `python3.9` interpreter or lower, I could reproduce this issue.

auria reviewed (commented):
LGTM.
